service/topology_coordinator.cc

```diff
@@ -740,6 +755,56 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
         }
     }
 
+    // If a node crashes after initiating gossip and before joining group0, it becomes orphan and 
+    // remains in gossip. This backgound fiber periodically checks for such nodes and purges those 
+    // entries from gossiper by commiting the node as left in raft.
+    future<> gossiper_orphan_remover_fiber() {
+        rtlogger.debug("start gossiper orphan remover fiber");
+
+        while (!_as.abort_requested()) {
+            try {
+                auto guard = co_await start_operation();
+                std::vector<canonical_mutation> updates;
+
+                _gossiper.for_each_endpoint_state([&](const gms::inet_address& addr, const gms::endpoint_state& eps) -> void {
+                    auto current_timestamp =
+                            std::chrono::duration_cast<std::chrono::seconds>(std::chrono::high_resolution_clock::now().time_since_epoch()).count();
+                    auto generation = eps.get_heart_beat_state().get_generation().value();
+                    static constexpr int32_t timeout = 60;
+                    auto host_id = eps.get_host_id();
+                    if (current_timestamp - generation > timeout && !_topo_sm._topology.contains(raft::server_id{host_id.id})) {
+                        topology_mutation_builder builder(guard.write_timestamp());
+                        builder.with_node(raft::server_id{host_id.id})
+                                .set("datacenter", get_application_state(host_id, addr, eps.get_application_state_map(), gms::application_state::DC))
+                                .set("rack", get_application_state(host_id, addr, eps.get_application_state_map(), gms::application_state::RACK))
+                                .set("release_version", version::release())
+                                .set("num_tokens", _db.get_config().join_ring() ? _db.get_config().num_tokens() : 0)
+                                .set("tokens_string", _db.get_config().join_ring() ? _db.get_config().initial_token() : sstring())
+                                .set("shard_count", smp::count)
+                                .set("ignore_msb", _db.get_config().murmur3_partitioner_ignore_msb_bits())
+                                .set("request_id", utils::UUID_gen::get_time_UUID())
+                                .set("cleanup_status", cleanup_status::clean)
+                                .set("node_state", node_state::left);
```

Comment:
Do we need to ban a node if it did not join group0?

Reply:
No we don't, but its just matter of nanoseconds that it would have crashed after joining group0(and would have got banned in that case).

secondly, purging via this mechanism, gives us a clean way and ensures race conditions dont hamper our intent. For example, we dont have tombstones for gossiper node removal, and this thing can lead to race conditions, but with raft, purges are easier since we have left node list as tombstone which maintains consistency.

Reply:
> No we don't, but its just matter of nanoseconds that it would have crashed after joining group0(and would have got banned in that case).

Sure. But we already have the code that bans nodes that joined group0. Here we need to write more code. If we can avoided it would be better.

> 
> secondly, purging via this mechanism, gives us a clean way and ensures race conditions dont hamper our intent. For example, we dont have tombstones for gossiper node removal, and this thing can lead to race conditions, but with raft, purges are easier since we have left node list as tombstone which maintains consistency.

I am not sure I understand the race.

Reply:
I described the race condition in the other comment: https://github.com/scylladb/scylladb/pull/21600#discussion_r1848320463

Reply:
This exact scenario cannot happen, but I see what you are looking for.  You are looking for a race where the entry is remove just before it is added to the topology by the `join_node_request_handler`. OK, I guess there is no harm to add it here. Do we need all those fields though?

Reply:
Yeah, it threw missing field errors without those fields. There Might be some checks somewhere in code

Reply:
These checks make a lot of sense for joining nodes that managed to go into topology through the happy path, but nodes that instantly go to `left` state (you introduce this type in this PR)? Maybe not necessary

Reply:
Right, its not absolutely necessary. Digged more on this, it seems like we need to make changes in system_keyspace.cc
```
        raft::server_id host_id{row.get_as<utils::UUID>("host_id")};
        auto datacenter = row.get_as<sstring>("datacenter");
        auto rack = row.get_as<sstring>("rack");
        auto release_version = row.get_as<sstring>("release_version");
        uint32_t num_tokens = row.get_as<int32_t>("num_tokens");
        sstring tokens_string = row.get_as<sstring>("tokens_string");
        size_t shard_count = row.get_as<int32_t>("shard_count");
        uint8_t ignore_msb = row.get_as<int32_t>("ignore_msb");
        sstring cleanup_status = row.get_as<sstring>("cleanup_status");
        utils::UUID request_id = row.get_as<utils::UUID>("request_id");
```

We need to make these fields optional here and even if we are dealing with left state path, we need to fill in replica state object:
```
        if (map) {
            map->emplace(host_id, service::replica_state{
                nstate, std::move(datacenter), std::move(rack), std::move(release_version),
                ring_slice, shard_count, ignore_msb, std::move(supported_features),
                service::cleanup_status_from_string(cleanup_status), request_id});
        }
```
Even if we make adjustments here, there are several other places in the call path which require these params. Even process_left_node uses this function:
```
    auto update_topology = [&] (locator::host_id id, std::optional<inet_address> ip, const replica_state& rs) {
        tmptr->update_topology(id, locator::endpoint_dc_rack{rs.datacenter, rs.rack},
                               to_topology_node_state(rs.state), rs.shard_count);
        if (ip) {
            tmptr->update_host_id(id, *ip);
        }
    };
```

I think even if we make changes everywhere, It will make code brittle and hard to maintain. If someone wants to add new field in replica state, it would be nightmare!

Reply:
Secondly, we are not compromising w/ dc and rack in my solution -- we are fetching those from gossiper, so they are accurate, rest of fields are from coordinator(or it can be dummy).

Reply:
Made the changes
Somehow managed to add num_tokens from gossiper itself.. used 0 for other places, to denote trivial entry.. also commented the same